### PR TITLE
fix(hooks): Reduce external request timeout to 10 s

### PR DESCRIPTION
**Pull request by timgl**
## Problem

Webhooks can take a long time to time out, causing issues elsewhere

## Changes

Reduce timeout to 10 seconds

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: plugin-server/src/worker/ingestion/hooks.ts [hunk 1]
```diff
@@ -240,10 +242,22 @@ export class HookCommander {
...
+                    team_id: event.teamId.toString(),
+                })
+            }, ms)
+            return ctrl.signal
+        }
         await fetch(webhookUrl, {
             method: 'POST',
             body: JSON.stringify(message, undefined, 4),
             headers: { 'Content-Type': 'application/json' },
+            signal: (AbortSignal as any).timeout(HOOK_TIMEOUT),
```
Comment (Twixes) [inline plugin-server/src/worker/ingestion/hooks.ts:260]: Why `as any`? It's a bad code smell

File: plugin-server/src/worker/ingestion/hooks.ts [hunk 2]
```diff
@@ -240,10 +242,22 @@ export class HookCommander {
                 text: messageMarkdown,
             }
         }
+        const statsd = this.statsd
+        ;(AbortSignal as any).timeout ??= function timeout(ms: number) {
```
Comment (Twixes) [inline plugin-server/src/worker/ingestion/hooks.ts:246]: Monkey-patching `AbortSignal` globally makes me feel dirty… Let's just use `addEventListener`  for monitoring (https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/abort_event#examples)

**Review by Twixes — CHANGES_REQUESTED** [review]
I wrote this a while ago, so took a look 👀

**Comment by Twixes** [comment]
100% in favor of reducing the timeout, my only point is we can do this without overriding built-ins :)

**Comment by Twixes** [comment]
Annoyingly we cannot quite _track_ those timeouts because all the options are blocked :/

1. Monkey-patching removes all safety guarantees and it's hard to say if it actually works.
2. `node-fetch` v2 (which we're using now) doesn't support Node 18's native `AbortSignal.timeout()` .
3. `node-fetch` v3 does support `AbortSignal.timeout()`, but it's an ES module, and the rest of the plugin server is CommonJS, and these don't mesh together. We _could_ migrate the plugin server to ES modules, but this is definitely beyond the scope of this PR (+ changing `node-fetch` version would affect plugins).
4. As we're on Node 18 now, we can also ditch `node-fetch` and use the native `fetch` API – this is probably best long-term, but it requires a slightly different approach to mocking in hook tests, which I don't wanna spend time on now.

So let's just use the `timeout` argument to `fetch`, which achieves our goal, the only downside being lack of timeout metrics.

**Comment by hazzadous** [comment]
> So let's just use the timeout argument to fetch, which achieves our goal, the only downside being lack of timeout metrics.

Better than nothing. btw what response do we get from `fetch` on timeouts? Is there nothing we can infer from this? i.e. would we get a rejection with error.name == "AbortError" or something similar?

**Comment by Twixes** [comment]
Ah of course, that's an `error.name == "AbortError"`. Theoretically that should be captured by Sentry, right?

**Review by Twixes — APPROVED** [review]
Looks like this should work, let's just get it in.

**Pull request merged**